Source/Core/VideoBackends/Vulkan/ShaderCache.cpp

```diff
@@ -0,0 +1,1377 @@
+      }
+    }
+  )";
+
+  std::string header = GetUtilityShaderHeader();
+
+  m_screen_quad_vertex_shader =
+      Util::CompileAndCreateVertexShader(header + SCREEN_QUAD_VERTEX_SHADER_SOURCE);
+  m_passthrough_vertex_shader =
+      Util::CompileAndCreateVertexShader(header + PASSTHROUGH_VERTEX_SHADER_SOURCE);
+  if (m_screen_quad_vertex_shader == VK_NULL_HANDLE ||
+      m_passthrough_vertex_shader == VK_NULL_HANDLE)
+  {
+    return false;
+  }
+
+  if (g_ActiveConfig.iStereoMode != STEREO_OFF && g_vulkan_context->SupportsGeometryShaders())
+  {
+    m_screen_quad_geometry_shader =
+        Util::CompileAndCreateGeometryShader(header + SCREEN_QUAD_GEOMETRY_SHADER_SOURCE);
+    m_passthrough_geometry_shader =
+        Util::CompileAndCreateGeometryShader(header + PASSTHROUGH_GEOMETRY_SHADER_SOURCE);
+    if (m_screen_quad_geometry_shader == VK_NULL_HANDLE ||
+        m_passthrough_geometry_shader == VK_NULL_HANDLE)
+    {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+void ShaderCache::DestroySharedShaders()
+{
+  auto DestroyShader = [this](VkShaderModule& shader) {
+    if (shader != VK_NULL_HANDLE)
+    {
+      vkDestroyShaderModule(g_vulkan_context->GetDevice(), shader, nullptr);
+      shader = VK_NULL_HANDLE;
+    }
+  };
+
+  DestroyShader(m_screen_quad_vertex_shader);
+  DestroyShader(m_passthrough_vertex_shader);
+  DestroyShader(m_screen_quad_geometry_shader);
+  DestroyShader(m_passthrough_geometry_shader);
+}
+
+void ShaderCache::CreateDummyPipeline(const UberShader::VertexShaderUid& vuid,
+                                      const GeometryShaderUid& guid,
+                                      const UberShader::PixelShaderUid& puid)
+{
+  PortableVertexDeclaration vertex_decl;
+  std::memset(&vertex_decl, 0, sizeof(vertex_decl));
+
+  PipelineInfo pinfo;
+  pinfo.vertex_format =
+      static_cast<const VertexFormat*>(VertexLoaderManager::GetUberVertexFormat(vertex_decl));
+  pinfo.pipeline_layout = g_object_cache->GetPipelineLayout(
+      g_ActiveConfig.bBBoxEnable && g_ActiveConfig.BBoxUseFragmentShaderImplementation() ?
+          PIPELINE_LAYOUT_BBOX :
+          PIPELINE_LAYOUT_STANDARD);
+  pinfo.vs = GetVertexUberShaderForUid(vuid);
+  pinfo.gs = (!guid.GetUidData()->IsPassthrough() && g_vulkan_context->SupportsGeometryShaders()) ?
+                 GetGeometryShaderForUid(guid) :
+                 VK_NULL_HANDLE;
+  pinfo.ps = GetPixelUberShaderForUid(puid);
+  pinfo.render_pass = FramebufferManager::GetInstance()->GetEFBLoadRenderPass();
+  pinfo.rasterization_state.bits = Util::GetNoCullRasterizationState().bits;
+  pinfo.depth_stencil_state.bits = Util::GetNoDepthTestingDepthStencilState().bits;
+  pinfo.blend_state.hex = Util::GetNoBlendingBlendState().hex;
+  pinfo.primitive_topology = VK_PRIMITIVE_TOPOLOGY_TRIANGLE_STRIP;
+  GetPipelineWithCacheResultAsync(pinfo);
+}
+
+void ShaderCache::PrecompileUberShaders()
+{
+  UberShader::EnumerateVertexShaderUids([&](const UberShader::VertexShaderUid& vuid) {
```

Comment:
D3D is generating vertex shaders alone, Vulkan/GL generate pipelines/programs. The pipelines need to have matching texgen counts across stages.